### PR TITLE
topdown/object: Rework `object.union_n` to use in-place merge algorithm.

### LINKED ISSUE #4985: object.union_n() is superlinear on the number of objects to merge
This is conceptually the same problem from #4979.

Suppose we are unioning together $n$ many objects of $k$ keys each. `topdown.builtinObjectUnionN()` will make $n$ many calls to `topdown.mergeWithOverwrite()`, implying $\Omega(1)$ calls to `ast.object.MergeWith()`. Though `MergeWith` appears linear at first glance, the calls are not independent from one another. The first call is merging two objects of $k$ keys each, the second call merges an object of $2k$ keys with an object of $k$ keys, and so on. Thus the actual work done by `MergeWith` in the inner loop ends up being: $O\left(2k + 3k + 4k + 5k + \dots\right) = O\left(k \left(\frac12 n (n+1) -1\right)\right) \approx O(kn^2)$.

This occurs in the optimal case where $\Omega(1) = 1$. If the objects contain complex nested objects as values rather than primitive types, then the actual running time could be substantially worse.

# Empirical Benchmarks

With the `object.union_n()` builtin:

```plain
$ cat slowunion/slowunion.rego
package slowunion

SIZE := 250
NSETS := 250

nums := numbers.range(0, NSETS)
sizes := numbers.range(0, SIZE)

sets[n] = x {
	nums[n]
	x := {sprintf("%d,%d", [n, i]): true | sizes[i]}
}

combined := object.union_n([s | s := sets[_]])
#combined := {k: true | s := sets[_]; s[k]}
$ time opa eval --profile --format pretty --bundle ./slowunion/ --import "data.slowunion" 'count(data.slowunion.combined)'
63001
+------------------------------+------------+
|            METRIC            |   VALUE    |
+------------------------------+------------+
| timer_rego_load_bundles_ns   | 533250     |
| timer_rego_module_compile_ns | 708000     |
| timer_rego_module_parse_ns   | 210625     |
| timer_rego_query_compile_ns  | 46625      |
| timer_rego_query_eval_ns     | 3028944250 |
| timer_rego_query_parse_ns    | 33875      |
+------------------------------+------------+
+-------------+----------+----------+-----------------------------+
|    TIME     | NUM EVAL | NUM REDO |          LOCATION           |
+-------------+----------+----------+-----------------------------+
| 2.88952125s | 4        | 254      | slowunion/slowunion.rego:14 |
| 110.7097ms  | 63503    | 126253   | slowunion/slowunion.rego:11 |
| 18.169001ms | 2        | 2        | data.slowunion.combined     |
| 154.583µs   | 3        | 3        | slowunion/slowunion.rego:6  |
| 114.531µs   | 1        | 251      | slowunion/slowunion.rego:10 |
| 91.833µs    | 3        | 3        | slowunion/slowunion.rego:7  |
| 4.458µs     | 1        | 1        | slowunion/slowunion.rego:4  |
| 2.042µs     | 1        | 1        | slowunion/slowunion.rego:3  |
+-------------+----------+----------+-----------------------------+

real	0m3.055s
user	0m4.560s
sys	0m0.153s
```

With pure Rego:

```
$ cat slowunion/slowunion.rego
package slowunion

SIZE := 250
NSETS := 250

nums := numbers.range(0, NSETS)
sizes := numbers.range(0, SIZE)

sets[n] = x {
	nums[n]
	x := {sprintf("%d,%d", [n, i]): true | sizes[i]}
}

#combined := object.union_n([s | s := sets[_]])
combined := {k: true | s := sets[_]; s[k]}
$ time opa eval --profile --format pretty --bundle ./slowunion/ --import "data.slowunion" 'count(data.slowunion.combined)'
63001
+------------------------------+-----------+
|            METRIC            |   VALUE   |
+------------------------------+-----------+
| timer_rego_load_bundles_ns   | 554167    |
| timer_rego_module_compile_ns | 710166    |
| timer_rego_module_parse_ns   | 154792    |
| timer_rego_query_compile_ns  | 48458     |
| timer_rego_query_eval_ns     | 194535167 |
| timer_rego_query_parse_ns    | 32042     |
+------------------------------+-----------+
+--------------+----------+----------+-----------------------------+
|     TIME     | NUM EVAL | NUM REDO |          LOCATION           |
+--------------+----------+----------+-----------------------------+
| 111.776641ms | 63503    | 126253   | slowunion/slowunion.rego:11 |
| 48.923215ms  | 254      | 63254    | slowunion/slowunion.rego:15 |
| 20.819749ms  | 2        | 2        | data.slowunion.combined     |
| 160.292µs    | 3        | 3        | slowunion/slowunion.rego:6  |
| 106.874µs    | 1        | 251      | slowunion/slowunion.rego:10 |
| 93.332µs     | 3        | 3        | slowunion/slowunion.rego:7  |
| 4.584µs      | 1        | 1        | slowunion/slowunion.rego:4  |
| 2.084µs      | 1        | 1        | slowunion/slowunion.rego:3  |
+--------------+----------+----------+-----------------------------+

real	0m0.218s
user	0m0.327s
sys	0m0.024s
```

System info:

```plain
$ uname -a
Darwin styra-macbook.local 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:37 PDT 2022; root:xnu-8020.121.3~4/RELEASE_ARM64_T6000 arm64
$ system_profiler SPSoftwareDataType SPHardwareDataType
Software:

    System Software Overview:

      System Version: macOS 12.4 (21F79)
      Kernel Version: Darwin 21.5.0
      Boot Volume: Macintosh HD
      Boot Mode: Normal
      Computer Name: styra-macbook
      User Name: Charles Daniels (cad)
      Secure Virtual Memory: Enabled
      System Integrity Protection: Enabled
      Time since boot: 5 days 21:41

Hardware:

    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro18,3
      Chip: Apple M1 Pro
      Total Number of Cores: 8 (6 performance and 2 efficiency)
      Memory: 32 GB
      System Firmware Version: 7459.121.3
      OS Loader Version: 7459.121.3
      Serial Number (system): REDACTED
      Hardware UUID: REDACTED
      Provisioning UDID: REDACTED
      Activation Lock Status: Disabled

$ opa version
Version: 0.43.0
Build Commit: d75bbdd
Build Timestamp: 2022-07-29T21:15:42Z
Build Hostname: Mac-1659129125033.local
Go Version: go1.18.4
Platform: darwin/arm64
WebAssembly: unavailable
```

Note that unlike in #4979, this is not an apples-to-apples comparison, because `object.union_n()` is not expressible in the general case (to my knowledge) using pure Rego. The fact that `object.union_n()` is slower than the given Rego code is not an "obvious" problem in of itself, it's the fact that the it grows far faster than linearly on the number of objects to merge.

Although the asymptotic analysis is sufficient on it's own to demonstrate this, I have implemented the following ad-hoc benchmark to demonstrate. Note that this needs a version of Bash new enough to support `EPOCHREALTIME`. Notice that though `nsets` varies with each run of the benchmark, `size` is fixed at 250 for all of them. This in the previous asymptotic analysis, $n$ is allowed to vary, and $k$ is fixed at 250.

```plain
$ cat slowunion/slowunion.rego
package slowunion

nums := numbers.range(0, input.nsets)
sizes := numbers.range(0, input.size)

sets[n] = x {
	nums[n]
	x := {sprintf("%d,%d", [n, i]): true | sizes[i]}
}

combined := object.union_n([s | s := sets[_]])
#combined := {k: true | s := sets[_]; s[k]}
$ rm out.csv
$ for n in 10 100 200 300 400 500 ; do START=$EPOCHREALTIME ; time opa eval --format pretty --bundle ./slowunion/ --import "data.slowunion" 'count(data.slowunion.combined) with input as {"nsets": '$n', "size": 250}' ; echo "$n,$START,$EPOCHREALTIME" >> out.csv ; done
$ cat out.csv
10,1659994724.161146,1659994724.199003
100,1659994724.199294,1659994724.692205
200,1659994724.692459,1659994726.621330
300,1659994726.621561,1659994731.005600
400,1659994731.005872,1659994738.928419
500,1659994738.928674,1659994751.609016
$ awk 'BEGIN{FS=","} {printf("n=%d, elapsed seconds=%f\n", $1, (1.0 * $3 - 1.0 * $2))}' < out.csv
n=10, elapsed seconds=0.037857
n=100, elapsed seconds=0.492911
n=200, elapsed seconds=1.928871
n=300, elapsed seconds=4.384039
n=400, elapsed seconds=7.922547
n=500, elapsed seconds=12.680342
```

As we can see, between n=100 and n=500 (factor of 5x increase in input size), the running time has gone from 0.49s to 12.68s, a 25x increase.

### PR DESCRIPTION
**Pull request by philipaconrad**
This PR fixes a performance regression for the `object.union_n` builtin, discovered in issue #4985.

The original logic for the builtin did pairwise `mergeWithOverwrite` calls between the input Objects, resulting in many wasted intermediate result Objects that were almost immediately discarded. The updated builtin uses a new algorithm to do a single pass across the input Objects, respecting the "last assignment wins, with merges" semantics of the original builtin implementation.

In the included benchmarks, this provides a 2x speed and 2-3x memory efficiency improvement over using a pure-Rego comprehension to do the same job, and a 6-30x improvement over the original implementation on all metrics as input Object arrays grow larger.

Fixes #4985

The 3x relevant benchmarks are shown below:

 - `object.union_n` on `main` branch
 - Pure Rego implementation (Object comprehension)
 - `object.union_n` on this PR

## Benchmarks

Benchmark of `object.union_n` on `main` branch:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkObjectUnionN$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkObjectUnionN/10x10-8         	    4965	    236921 ns/op	   79700 B/op	    1441 allocs/op
BenchmarkObjectUnionN/10x100-8        	     526	   2305226 ns/op	  759100 B/op	   11524 allocs/op
BenchmarkObjectUnionN/10x250-8        	     128	  11340810 ns/op	 1946964 B/op	   28243 allocs/op
BenchmarkObjectUnionN/100x10-8        	      34	  32815013 ns/op	 7083399 B/op	  104789 allocs/op
BenchmarkObjectUnionN/100x100-8       	       4	 267571459 ns/op	74272772 B/op	 1025478 allocs/op
BenchmarkObjectUnionN/100x250-8       	       2	 778593630 ns/op	182422912 B/op	 2562647 allocs/op
BenchmarkObjectUnionN/250x10-8        	       7	 171022894 ns/op	44684044 B/op	  641897 allocs/op
BenchmarkObjectUnionN/250x100-8       	       1	1771797807 ns/op	453937248 B/op	 6368408 allocs/op
BenchmarkObjectUnionN/250x250-8       	       1	4780766324 ns/op	1205083376 B/op	15958999 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	20.584s
```

Benchmark of Pure Rego implementation:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkObjectUnionNSlow$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkObjectUnionNSlow/10x10-8         	    9382	    130836 ns/op	   40207 B/op	     818 allocs/op
BenchmarkObjectUnionNSlow/10x100-8        	    1122	   1202061 ns/op	  323580 B/op	    5348 allocs/op
BenchmarkObjectUnionNSlow/10x250-8        	     535	   2209046 ns/op	  744128 B/op	   12876 allocs/op
BenchmarkObjectUnionNSlow/100x10-8        	    1210	   1025966 ns/op	  368944 B/op	    6788 allocs/op
BenchmarkObjectUnionNSlow/100x100-8       	     100	  10117234 ns/op	 3093754 B/op	   51983 allocs/op
BenchmarkObjectUnionNSlow/100x250-8       	      42	  26226170 ns/op	 7340888 B/op	  127626 allocs/op
BenchmarkObjectUnionNSlow/250x10-8        	     452	   2805731 ns/op	  865134 B/op	   16867 allocs/op
BenchmarkObjectUnionNSlow/250x100-8       	      46	  25466460 ns/op	 7417977 B/op	  130189 allocs/op
BenchmarkObjectUnionNSlow/250x250-8       	      18	  69078588 ns/op	20508967 B/op	  319094 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	15.404s
```

Benchmark of `object.union_n` on this PR:
```
Running tool: /usr/local/go/bin/go test -benchmem -run=^$ -bench ^BenchmarkObjectUnionN$ github.com/open-policy-agent/opa/topdown

goos: linux
goarch: amd64
pkg: github.com/open-policy-agent/opa/topdown
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkObjectUnionN/10x10-8         	   25534	     44702 ns/op	   19468 B/op	     374 allocs/op
BenchmarkObjectUnionN/10x100-8        	    4594	    458595 ns/op	  158814 B/op	    2203 allocs/op
BenchmarkObjectUnionN/10x250-8        	    1677	    848999 ns/op	  339319 B/op	    5232 allocs/op
BenchmarkObjectUnionN/100x10-8        	    2611	    570072 ns/op	  161686 B/op	    2293 allocs/op
BenchmarkObjectUnionN/100x100-8       	     340	   3914881 ns/op	 1446257 B/op	   20485 allocs/op
BenchmarkObjectUnionN/100x250-8       	     153	   7588882 ns/op	 3294123 B/op	   51133 allocs/op
BenchmarkObjectUnionN/250x10-8        	    1686	    636962 ns/op	  347029 B/op	    5472 allocs/op
BenchmarkObjectUnionN/250x100-8       	     163	   7328428 ns/op	 3298907 B/op	   51283 allocs/op
BenchmarkObjectUnionN/250x250-8       	      57	  20065940 ns/op	10388723 B/op	  127676 allocs/op
PASS
ok  	github.com/open-policy-agent/opa/topdown	17.998s
```

## Summary of Results

As mentioned above, and as shown in the benchmark results, this PR's changes provide dramatic performance improvements on all metrics, and result in the builtin having around 2-3x better performance overall than a pure Rego equivalent.

2022-09-12_object-union-n-optimization-v2

### CODE REVIEW

**Comment by philipaconrad** [comment]
While none of our current tests are erroring, I may have found an edge case for the algorithm as it is right now. I'll write up some test cases for it, and if my suspicions are founded, I'll go back and re-benchmark after patching.

File: topdown/object.go [hunk 1]
```diff
@@ -32,21 +32,30 @@ func builtinObjectUnionN(_ BuiltinContext, operands []*ast.Term, iter func(*ast.
...
+	// give us the "last assignment wins, with merges" behavior we want.
+	// However, if a non-object overwrites an object value anywhere in the
+	// chain of assignments for a key, we have to "freeze" that key to
+	// prevent accidentally picking up nested objects that could merge with
+	// it from earlier in the input array.
+	// Example:
+	//   Input: [{"a": {"b": 2}}, {"a": 4}, {"a": {"c": 3}}]
+	//   Want Output: {"a": {"c": 3}}
+	result := ast.NewObject()
+	frozenKeys := map[*ast.Term]struct{}{}
```
Comment (philipaconrad) [inline topdown/object.go:46]: We need a way to know when a key has been "frozen" by an assignment, such that any objects ahead of it are not allowed to be used for recursive merging.

The above example is illustrative of the issue. The middle assignment "blocks" the first assignment of the key `"a"`, so we need some way to ensure we don't accidentally try to merge the value `{"b": 2}` into the final result object.

Luckily, because pointers to distinct keys in the result object will be unique, we can create a set of the keys in the result Object that should be "frozen" by [using a map as a set-like data structure](https://go.dev/blog/maps#exploiting-zero-values). This set of frozen keys is passed down into the recursive merging function.

File: topdown/object.go [hunk 2]
```diff
@@ -173,6 +182,32 @@ func mergeWithOverwrite(objA, objB ast.Object) ast.Object {
...
+		// Both are objects? Merge recursively.
+		if ok1 && ok2 {
+			// Check to make sure that this key isn't frozen before merging.
+			if _, ok := frozenKeys[v2]; !ok {
+				mergewithOverwriteInPlace(originalValueObj, updateValueObj, frozenKeys)
+			}
+		} else {
+			// Else, original value wins. Freeze the key.
+			frozenKeys[v2] = struct{}{}
+		}
```
Comment (philipaconrad) [inline topdown/object.go:207]: If we see two objects, and the key isn't in the "frozen" set, we can merge them recursively.

However, if one of the values did not correspond to an object type, we should freeze the key because we've found a "blocking" assignment, such that it would have overwritten everything that came before it.

File: topdown/object.go [hunk 3]
```diff
@@ -32,21 +32,30 @@ func builtinObjectUnionN(_ BuiltinContext, operands []*ast.Term, iter func(*ast.
...
+	// chain of assignments for a key, we have to "freeze" that key to
+	// prevent accidentally picking up nested objects that could merge with
+	// it from earlier in the input array.
+	// Example:
+	//   Input: [{"a": {"b": 2}}, {"a": 4}, {"a": {"c": 3}}]
+	//   Want Output: {"a": {"c": 3}}
+	result := ast.NewObject()
+	frozenKeys := map[*ast.Term]struct{}{}
+	for i := arr.Len() - 1; i >= 0; i-- {
+		o, err := builtins.ObjectOperand(arr.Elem(i).Value, 1)
```
Comment (srenatus) [inline topdown/object.go:48]: [nit] not new code, but I think this error will be slight off the mark. If it we're passing `[{"foo":"2"}, 1]`, it'll complain about operand 1 not being an object, I think.

**Review by srenatus — APPROVED** [review]

File: topdown/object.go (hunk 3 above)
Comment (philipaconrad) [reply topdown/object.go:48]: As I understand it, this builtin is supposed to take in an array of objects. The example input `[{"foo":"2"}, 1]` indeed should be rejected, right?

Comment (srenatus) [reply topdown/object.go:48]: It should but the error will say something about requiring an object operand 1, not an array of objects.

Comment (philipaconrad) [reply topdown/object.go:48]: Ah, I see. I'll see what I can do to render a better error message there.

Comment (philipaconrad) [reply topdown/object.go:48]: `builtins.NewOperandElementErr` looks to be pretty close to what we want on a failed type assertion here, I think?

Comment (srenatus) [reply topdown/object.go]: 💡 I wasn't even aware that this existed. Great. Yes!

**Review by srenatus — APPROVED** [review]
Thanks!
